### ISSUE TITLE
bytes API

### ISSUE THREAD

**Issue by CeleritasCelery** [cessen/str_indices#16]
Most (all?) of the functions in this crate don't require any of the invariants of a `str`. They all take in the `&str` and then immediately cast it to bytes.  There are some use cases where it would be beneficial to operate over bytes instead of strings. This can because you don't want to run UTF-8 validation on some stream of bytes, or may not care if it is valid UTF-8. Or in my case I need the ability to take fixed sized chunks of memory and count things like starts or characters or linefeeds. The beginning and end of the chunk may not be a char boundary. This would allow this crate to be used in more places where currently I am using `bytecount` because it [operates on byte slices](https://docs.rs/bytecount/0.6.3/bytecount/fn.num_chars.html). 

Would you be willing to add a bytes API to this crate? I would do the implementation work if you provide the direction you want to take. I can think of two ways to handle this:

### 1 - change the public methods to take `impl AsRef<[u8]>`
This has the advantage of making all the existing call sites work correctly (since `&str` implements that trait). I think this would technically make it non-breaking change. But it also allows passing anything that can `AsRef` to a byte slice, which would include `String`, `Vec<u8>`, and of course `&[u8]`.

### 2 - add a seperate `_byte` version of each function
This would allow the current functions to remain unchanged (and concrete) while still exposing a byte API. These could even be put in a separate module like the [regex crate does](https://docs.rs/regex/1.7.1/regex/bytes/index.html). I am less a fan of this because it makes the public API surface larger.

**Comment by cessen** [comment]
In theory I think this is a good idea, yeah.  But the tricky part isn't choosing the surface-level API (I'm fine with either of the approaches you outlined).  It's in determining what semantics we can/should guarantee with non-utf8 byte slices.

For example, the current implementation of `chars::count()` works by counting the bytes that don't match the required pattern for the start byte of a code point, and then subtracting that from the total byte count.  This is really fast, and switching to another approach would likely harm performance.  But for non-utf8 conformant byte slices this could lead to some pretty unintuitive behaviors, where e.g. things that aren't full code points get counted as such.

Additionally, even if we're okay with some weird semantics, I'm concerned about *committing* to those semantics because we may discover other faster approaches in the future that work identically for `&str` but change the semantics for arbitrary `&[u8]`.

One possibility would be to say, "The semantics are only guaranteed for full, valid utf8 data."  But that wouldn't help your use case where the beginning and end of the chunk may not be a char boundary.

**Comment by CeleritasCelery** [comment]
I have been thinking about your question regarding semantics, and I feel like I have a better answer. The functions in this crate are broken into two "types": counting and indexing. Let's start with counting.

### counting

The counting functions are the ones that really benefit from operating over bytes slices because it is a [monoid homomorphism](https://xi-editor.io/docs/rope_science_02.html). This means that you do interesting things like compute it in parallel or distribute it in fixed sized chunks. Essentially this means that this property holds true.

```rust
count(slice[a..c]) == count(slice[a..b]) + count(slice[b..c])
```

My desire for bytes API semantics would be to guarantee the homomorphic property even if you don't slice it on char boundaries. Conveniently this already how the code works today (though you could still change the way the code handles characters and uphold this property). This would let you grab arbitrary chunks of data (maybe from a file reader or stream) and count something and know that you will get the same answer without trying to handle all the edge cases. 

Now I think it would be worth clarifying that this is *only* valid for slices of well-formed UTF-8. If you passed in random bytes, you have no guarantees.

### indexing

I have been thinking more about the indexing API's and I don't think they would benefit from using bytes. For one thing, There are some weird semantics when you split characters. This is because these methods are relying on certain invariants of the `str` type. For example consider this code:

```rust
let s = "せ";
let b = s.as_bytes()[..2]; // split the first char
let i = chars::from_byte_idx(b, 2);
// i = 1
```
This will return that the byte indexing of 2 is char index 1, which is not true. It is still part of char 0, but the code is assuming that the end of the slice is a complete character. However if you slice on bytes you can no longer rely on that property. It also does not really compose with other methods when you split characters. Here is another example of problematic edge cases with indexing:

```rust

let s = "せか";
let b1 = s.as_bytes()[..4]; // split the second char
let b2 = s.as_bytes()[4..];
let count = chars::count(b1); // returns 2
let idx = chars::from_byte_idx(b2, 1);
let full_idx = chars::from_byte_idx(s, b1.len() + 1);
let compose_idx = count + idx;
```

`full_idx` will return 1 but `compose_idx` will return 2. But you only get this mismatch when you split characters. It works correctly with `str`. 

### conclusion

Due to all the semantic issue with indexing partial characters, I don't think that would be a good thing for this library to support. But I do think byte slices with the counting API's would be very valuable. This would eliminate a need for the `bytecount` crate but also provide a reliable guarantee.

**Comment by cessen** [comment]
Thanks for the additional thoughts, and the thoughtful writeup!

However, I've been thinking about this more, and I don't think I want to support byte slice APIs in `str_indices` after all.  My rationale is threefold:

- This crate explicitly targets only valid, well-formed utf8 data.  Handling arbitrary byte sequences is out of scope.
- So that only leaves the split-mid-codepoint use case that you outlined.  However, ensuring that splits only happen at `char` boundaries is both fast and not especially difficult, so that seems like a pretty minor burden to place on client code.
- The benefit of imposing that minor burden is that the whole question of what semantics arbitrary byte slices should have simply disappears.  And as a bonus, the API surface area is both smaller and enforces its invariants statically.

In short, I pretty much only see drawbacks to adding byte slice support, with very little practical benefit within `str_indices`' intended scope.

> This would let you grab arbitrary chunks of data (maybe from a file reader or stream) and count something and know that you will get the same answer without trying to handle all the edge cases.

I think the file reader is a really good example of where this isn't nearly as useful as it sounds.  If you're reading utf8 data from a file, you need to validate that data at some point regardless, which effectively requires finding valid char boundaries.  And at that point, you already have the boundaries needed to pass valid str slices anyway.

I can see better the use case of using multiple threads to e.g. count the chars in a really large string.  Being able to split at arbitrary byte indices certainly makes that more convenient, because you don't have to find char boundaries when splitting.  But that's mainly a convenience thing.  It's not hard to find the next char boundary after an arbitrary byte index.

Granted, if you want to use something like `rayon` to handle the threads, then you'd need to wrap the chunking logic in an iterator, which is a bit more work.  But it's still not too bad.

So in short, unless there's a compelling (and in scope) use case that I'm missing, I think I'll leave `str_indices` API's as `str`-only.
